File: extension/systemTray.js
/**
 * System Tray Integration module for CmdBar KDE Plasma Integration.
 * Implements StatusNotifierItem D-Bus interface (org.kde.StatusNotifierItem / org.freedesktop.StatusNotifierItem)
 * and freedesktop notifications (org.freedesktop.Notifications).
 */

export class SystemTrayManager {
  constructor(options = {}) {
    this.id = options.id || "cmdbar-system-tray";
    this.title = options.title || "CmdBar Quick Menu";
    this.iconName = options.iconName || "utilities-terminal";
    this.status = options.status || "Active"; // Active, Passive, NeedsAttention
    this.tooltip = options.tooltip || "CmdBar Command Palette";
    this.badgeCount = options.badgeCount || 0;
    this._menuItems = options.menuItems || [
      { id: "toggle", label: "Toggle CmdBar Menu", action: null },
      { id: "exec", label: "Run Command...", action: null },
      { id: "prefs", label: "Preferences", action: null },
      { id: "quit", label: "Quit", action: null },
    ];
    this._isVisible = true;
    this._notifications = [];
  }

  /**
   * Updates system tray icon status or badge count.
   * @param {string} status 'Active' | 'Passive' | 'NeedsAttention'
   * @param {number} badgeCount
   */
  updateStatus(status = "Active", badgeCount = 0) {
    this.status = status;
    this.badgeCount = Math.max(0, badgeCount);
    return { status: this.status, badgeCount: this.badgeCount };
  }

  /**
   * Sets custom tooltip text for system tray icon.
   * @param {string} text
   */
  setTooltip(text) {
    if (text && typeof text === "string") {
      this.tooltip = text.trim();
    }
    return this.tooltip;
  }

  /**
   * Returns list of context menu items.
   * @returns {Array<{id: string, label: string}>}
   */
  getMenuItems() {
    return [...this._menuItems];
  }

  /**
   * Simulates clicking a menu item by ID.
   * @param {string} itemId
   */
  triggerMenuItem(itemId) {
    const item = this._menuItems.find((m) => m.id === itemId);
    if (item && typeof item.action === "function") {
      item.action(itemId);
      return true;
    }
    return false;
  }

  /**
   * Registers a handler callback for menu item.
   * @param {string} itemId
   * @param {Function} callback
   */
  onMenuItem(itemId, callback) {
    const item = this._menuItems.find((m) => m.id === itemId);
    if (item) {
      item.action = callback;
      return true;
    }
    this._menuItems.push({ id: itemId, label: itemId, action: callback });
    return true;
  }

  /**
   * Sends desktop notification over org.freedesktop.Notifications D-Bus service.
   * @param {string} title Notification header
   * @param {string} body Notification message body
   * @param {string} icon Icon name (default: utilities-terminal)
   * @returns {Promise<number>} Notification ID
   */
  async sendNotification(title, body, icon = "utilities-terminal") {
    const notifObj = {
      id: this._notifications.length + 1,
      title: title || "CmdBar",
      body: body || "",
      icon: icon || this.iconName,
      timestamp: Date.now(),
    };
    this._notifications.push(notifObj);

    if (typeof globalThis.Gio !== "undefined" && globalThis.Gio.DBus) {
      try {
        const bus = globalThis.Gio.DBus.session;
        await new Promise((resolve) => {
          bus.call(
            "org.freedesktop.Notifications",
            "/org/freedesktop/Notifications",
            "org.freedesktop.Notifications",
            "Notify",
            new globalThis.GLib.Variant("(susssasa{sv}i)", [
              "CmdBar",
              0,
              notifObj.icon,
              notifObj.title,
              notifObj.body,
              [],
              {},
              3000,
            ]),
            new globalThis.GLib.VariantType("(u)"),
            globalThis.Gio.DBusCallFlags.NONE,
            -1,
            null,
            (connection, res) => {
              try {
                const val = bus.call_finish(res);
                if (val) notifObj.id = val.recursiveUnpack()[0];
              } catch (e) {}
              resolve(true);
            }
          );
        });
      } catch (e) {}
    }

    return notifObj.id;
  }

  /**
   * Toggles system tray icon visibility.
   * @param {boolean} visible
   */
  setVisible(visible) {
    this._isVisible = Boolean(visible);
    return this._isVisible;
  }

  get isVisible() {
    return this._isVisible;
  }
}

File: companion/kde_plasma/system_tray.py
"""
System Tray Manager for CmdBar Python Companion in KDE Plasma.
Implements org.kde.StatusNotifierItem / org.freedesktop.StatusNotifierItem D-Bus interface
and org.freedesktop.Notifications integration.
"""

import time
import subprocess
from typing import Dict, Any, List, Optional, Callable


class SystemTrayManager:
    def __init__(self, id_str: str = "cmdbar-system-tray", title: str = "CmdBar Quick Menu", icon_name: str = "utilities-terminal"):
        self.id = id_str
        self.title = title
        self.icon_name = icon_name
        self.status = "Active"  # Active, Passive, NeedsAttention
        self.tooltip = "CmdBar Command Palette"
        self.badge_count = 0
        self.is_visible = True
        self._menu_items = [
            {"id": "toggle", "label": "Toggle CmdBar Menu", "action": None},
            {"id": "exec", "label": "Run Command...", "action": None},
            {"id": "prefs", "label": "Preferences", "action": None},
            {"id": "quit", "label": "Quit", "action": None},
        ]
        self._notifications: List[Dict[str, Any]] = []

    def update_status(self, status: str = "Active", badge_count: int = 0) -> Dict[str, Any]:
        """
        Updates status notifier item status ('Active', 'Passive', 'NeedsAttention') and badge count.
        """
        if status in ("Active", "Passive", "NeedsAttention"):
            self.status = status
        self.badge_count = max(0, badge_count)
        return {"status": self.status, "badge_count": self.badge_count}

    def set_tooltip(self, text: str) -> str:
        if text and isinstance(text, str):
            self.tooltip = text.strip()
        return self.tooltip

    def get_menu_items(self) -> List[Dict[str, Any]]:
        return list(self._menu_items)

    def trigger_menu_item(self, item_id: str) -> bool:
        for item in self._menu_items:
            if item["id"] == item_id:
                if item["action"] and callable(item["action"]):
                    item["action"](item_id)
                    return True
        return False

    def on_menu_item(self, item_id: str, callback: Callable[[str], None]) -> bool:
        for item in self._menu_items:
            if item["id"] == item_id:
                item["action"] = callback
                return True
        self._menu_items.append({"id": item_id, "label": item_id, "action": callback})
        return True

    def send_notification(self, title: str, body: str, icon: Optional[str] = None) -> int:
        """
        Sends desktop notification using notify-send or D-Bus org.freedesktop.Notifications.
        """
        notif_id = len(self._notifications) + 1
        icon_str = icon or self.icon_name
        notif_obj = {
            "id": notif_id,
            "title": title or "CmdBar",
            "body": body or "",
            "icon": icon_str,
            "timestamp": time.time(),
        }
        self._notifications.append(notif_obj)

        try:
            subprocess.run([
                "notify-send",
                "-i", icon_str,
                title or "CmdBar",
                body or ""
            ], capture_output=True, text=True, timeout=2)
        except Exception:
            pass

        return notif_id

    def set_visible(self, visible: bool) -> bool:
        self.is_visible = bool(visible)
        return self.is_visible
